classes/task/process_course_task.php

```diff
@@ -31,15 +31,45 @@ class process_course_task extends \core\task\adhoc_task {
      * Execute the task.
      */
     public function execute() {
-        $oscourseid = $this->get_custom_data();
-        !PHPUNIT_TEST ? mtrace('[INFO] Process course task started') : false;
-        $handler = null;
-        $handler = new opensesame_handler();
-        $success = $handler->process_single_os_course($oscourseid);
-        !PHPUNIT_TEST ? mtrace('[INFO] Process course task finished') : false;
-        if (!$success) {
-            self::queue_task($oscourseid);
+        global $DB;
+        $failcount = get_config('tool_opensesame', 'process_course_task_fails_count');
+        $maxfails = get_config('tool_opensesame', 'max_consecutive_fails');
+        $maxfails = !empty($maxfails) ? $maxfails : 5;
+        if ($maxfails < $failcount) {
+            $oscourseid = $this->get_custom_data();
+            !PHPUNIT_TEST ? mtrace('[INFO] Process course task started') : false;
+            $handler = null;
+            $handler = new opensesame_handler();
+            $success = $handler->process_single_os_course($oscourseid);
+            !PHPUNIT_TEST ? mtrace('[INFO] Process course task finished') : false;
+            if ($success) {
+                // Restart count.
+                set_config('process_course_task_fails_count', 0, 'tool_opensesame');
+            } else {
+                $failcount = !empty($failcount) ? $failcount + 1 : 1;
+                set_config('process_course_task_fails_count', $failcount, 'tool_opensesame');
+                self::queue_task($oscourseid);
+                if ($failcount >= $maxfails) {
+                    !PHPUNIT_TEST ? mtrace('PURGING OPENSESAME TASKS DUE CONSECUTIVE FAILS') : false;
+                    // Let's clean adhoc task table.
+                    $adhoctasks = $DB->get_recordset('task_adhoc', ['component' => 'tool_opensesame']);
+                    foreach($adhoctasks as $adhoctask) {
+                        $DB->delete_records('task_adhoc', ['id' => $adhoctask->id]);
```

Comment:
I'm wondering what will happen if you have a lot of tasks executing simultaneously and they all start trying to delete the same rows from the database.  If a task calls $DB->delete_records on a task id that does not exist, does that function throw an exception?  If it throws an exception, then you could end up in a situation where all of the tasks get an exception which means they exit early and the status for the queued courses does not get set back to 'retrieved' (see below)

Reply:
You are right I'll add a lock here too, I added a look when updating the count but not here.